docs/modules/password_auth_provider_callbacks.md

```diff
@@ -0,0 +1,155 @@
+
+A dict mapping from tuples of a login type identifier (such as `m.login.password`) and a
+tuple of field names (such as `("password", "secret_thing")`) to authentication checking
+callbacks.
+
+The login type and field names are the things that should be provided by the user in their
+request to the `/login` API. 
+
+For example, if a module implements authentication checkers for two different login types: 
+-  `com.example.custom_login` 
+    - Expects `secret1` and `secret2` fields to be sent to `/login`
+    - Is checked by the method: `self.custom_login_check`
+- `m.login.password`
+    - Expects a `password` field to be sent to `/login`
+    - Is checked by the method: `self.password_checker` 
+    
+Then it should register the following dict:
+
+```python
+{ 
+    ("com.example.custom_login", ("secret1", "secret2")): self.check_custom_login,
+    ("m.login.password", ("password",)): self.password_checker,
+}
+```
+
+An authentication checking method should be of the following form:
+
+```python
+async def check_auth(
+    username: str,
+    login_type: str,
+    login_dict: "synapse.module_api.JsonDict",
+) -> Optional[
+    Tuple[
+        str, 
+        Optional[Callable[["synapse.module_api.LoginResponse"], Awaitable[None]]]
+    ]
+]
+```
+
+It is passed the user field provided by the client (which might not be in `@username:server` form), 
+the login type, and a dictionary of login secrets passed by the client.
+
+If the authentication is successful, the module must return the user's Matrix ID (e.g. @alice:example.com) and optionally a callback to be called with the response to the `/login` request. If the module doesn't wish to return a callback, it must return None instead.
+
+If the authentication is unsuccessful, the module must return None.
+
+### `check_3pid_auth`
+
+```python
+async def check_3pid_auth(
+    medium: str, 
+    address: str,
+    password: str,
+) -> Optional[
+    Tuple[
+        str, 
+        Optional[Callable[["synapse.module_api.LoginResponse"], Awaitable[None]]]
+    ]
+]
+```
+
+Called when a user attempts to register or log in with a third party identifier,
+such as email. It is passed the medium (eg. "email"), an address (eg. "jdoe@example.com")
+and the user's password.
+
+If the authentication is successful, the module must return the user's Matrix ID (e.g. @alice:example.com) and optionally a callback to be called with the response to the `/login` request. If the module doesn't wish to return a callback, it must return None instead.
+
+If the authentication is unsuccessful, the module must return None.
+
+### `on_logged_out`
+
+```python
+async def on_logged_out(
+    user_id: str,
+    device_id: Optional[str],
+    access_token: str
+) -> None
+``` 
+Called during a logout request for a user. It is passed the qualified user ID, the ID of the
+deactivated device (if any: access tokens are occasionally created without an associated
+device ID), and the (now deactivated) access token.
+
+## Example
```

Comment:
Could you add a few lines to explain what this example does please? So the reader knows what's going on, like eg https://matrix-org.github.io/synapse/develop/modules/spam_checker_callbacks.html#example